### PR TITLE
Connectivity events and API

### PR DESCRIPTION
**Pull request by steils**
Connectivity status and events support for zenoh-pico, introduced in https://github.com/eclipse-zenoh/zenoh/pull/2301

- Add connectivity info and listener APIs for transports/links.
- Implement transport/link connectivity events on peer connect/disconnect.
- Enable this feature via Z_FEATURE_CONNECTIVITY (requires
  Z_FEATURE_UNSTABLE_API).
- Add test z_api_connectivity_test.
- Support the feature in z_info example.

### Related Issues

https://github.com/eclipse-zenoh/zenoh-pico/issues/333
https://github.com/eclipse-zenoh/zenoh-pico/issues/369
https://github.com/eclipse-zenoh/zenoh-pico/issues/1007

---
## 🏷️ Label-Based Checklist

Based on the labels applied to this PR, please complete these additional requirements:

**Labels:** `new feature`

## 🆕 New Feature Requirements

Since this PR adds a new feature:

- [x] **Feature scope documented** - Clear description of what the feature does and why it's needed
- [x] **Minimum necessary code** - Implementation is as simple as possible, doesn't overcomplicate the system
- [x] **New APIs well-designed** - Public APIs are intuitive, consistent with existing APIs
- [x] **Comprehensive tests** - All functionality is tested (happy path + edge cases + error cases)
- [x] **Examples provided** - Usage examples in code comments or separate example files
- [x] **Documentation added** - New docs explaining the feature, its use cases, and API
- [x] **Feature flag considered** - Can the feature be enabled/disabled for gradual rollout?
- [x] **Performance impact assessed** - Memory, CPU, storage implications measured
- [x] **Integration tested** - Feature works with existing features

**Consider:** Can this feature be split into smaller, incremental PRs?

**Instructions:**
1. Check off items as you complete them (change `- [ ]` to `- [x]`)
2. The PR checklist CI will verify these are completed

*This checklist updates automatically when labels change, but preserves your checked boxes.*

### CODE REVIEW

File: docs/api.rst [hunk 1]
```diff
@@ -1764,6 +1900,7 @@ Enabling the Admin Space
 
 The Admin Space is an **optional feature** and must be explicitly enabled at
 build time by defining the ``Z_FEATURE_ADMIN_SPACE`` configuration flag.
+It requires both ``Z_FEATURE_UNSTABLE_API`` and ``Z_FEATURE_CONNECTIVITY``.
```
Comment (gmartin82) [inline docs/api.rst]: The admin space feature should have no dependency on the connectivity feature.

If extra information can be contributed to the admin space when the connectivity feature is enabled, then only the provision of that additional information should be dependent on the connectivity feature.

File: docs/config.rst [hunk 2]
```diff
@@ -165,4 +166,4 @@ The following options are here to reduce binary sizes for users that don't need
 * `Z_FEATURE_LINK_SERIAL`: (DEFAULT: OFF) Toggle compilation of Serial link support.
 * `Z_FEATURE_LINK_SERIAL_USB`: (DEFAULT: OFF) Toggle compilation of Serial USB link support.
 * `Z_FEATURE_LINK_TLS`: (DEFAULT: OFF) Toggle compilation of TLS support.
-* `Z_FEATURE_ADMIN_SPACE`: (DEFAULT: OFF) Toggle compilation of admin space API functions.
```
Comment (gmartin82) [inline docs/config.rst:168]: See previous comment. Admin space should not depend on the connectivity feature.

File: src/api/admin_space.c [hunk 3]
```diff
@@ -241,17 +386,18 @@ void _ze_admin_space_reply_clear(_ze_admin_space_reply_t *reply) {
     _ze_admin_space_reply_null(reply);
 }
 
-static z_result_t _ze_admin_space_add_reply(_z_json_encoder_t *je, const z_loaned_keyexpr_t *ke,
-                                            _ze_admin_space_reply_list_t **replies) {
+static z_result_t _ze_admin_space_add_reply_payload(const z_loaned_keyexpr_t *ke, z_owned_bytes_t *payload,
+                                                    _ze_admin_space_reply_list_t **replies) {
     _ze_admin_space_reply_t *reply = z_malloc(sizeof(_ze_admin_space_reply_t));
     if (reply == NULL) {
+        z_bytes_drop(z_bytes_move(payload));
```
Comment (gmartin82) [inline src/api/admin_space.c]: If this function can consume bytes, then it should accept a z_moved_bytes_t to make this explicit.

File: src/api/admin_space.c [hunk 4]
```diff
@@ -93,141 +111,268 @@ static z_result_t _ze_admin_space_transport_link_ke(z_owned_keyexpr_t *ke, const
...
-            return _z_json_encoder_write_string(je, "peer");
-        case Z_WHATAMI_CLIENT:
-            return _z_json_encoder_write_string(je, "client");
-        default:
-            return _Z_ERR_INVALID;
+static inline const char *_ze_admin_space_transport_kind_from_session(const _z_session_t *session, bool is_multicast) {
+    if (session != NULL && session->_tp._type == _Z_TRANSPORT_RAWETH_TYPE) {
+        return _Z_KEYEXPR_TRANSPORT_RAWETH;
     }
+    return is_multicast ? _Z_KEYEXPR_TRANSPORT_MULTICAST : _Z_KEYEXPR_TRANSPORT_UNICAST;
```
Comment (gmartin82) [inline src/api/admin_space.c]: Why are we determining the type in this way? _type is an enum value.

File: src/api/admin_space.c [hunk 5]
```diff
@@ -93,141 +111,268 @@ static z_result_t _ze_admin_space_transport_link_ke(z_owned_keyexpr_t *ke, const
...
+    if (z_link_event_kind(event) == Z_SAMPLE_KIND_PUT) {
+        z_owned_bytes_t payload;
+        ret = _ze_admin_space_encode_link_event_payload(&payload, link);
+        if (ret == _Z_RES_OK) {
+            z_put_options_t opts;
+            z_put_options_default(&opts);
+            z_owned_encoding_t encoding;
+            if (z_encoding_clone(&encoding, z_encoding_application_json()) == _Z_RES_OK) {
+                opts.encoding = z_encoding_move(&encoding);
+                ret = z_put(&session_rc, z_keyexpr_loan(&ke), z_bytes_move(&payload), &opts);
```
Comment (gmartin82) [inline src/api/admin_space.c]: This should be gated on `Z_FEATURE_PUBLICATION`. Admin space should still function without this feature if `Z_FEATURE_QUERYABLE` is enabled.

File: CMakeLists.txt [hunk 6]
```diff
@@ -242,6 +242,7 @@ set(Z_TRANSPORT_LEASE_EXPIRE_FACTOR 3 CACHE STRING "Default session lease expire
 set(ZP_PERIODIC_SCHEDULER_MAX_TASKS 64 CACHE STRING "Maximum number of tasks in the periodic scheduler")
 
 set(Z_FEATURE_UNSTABLE_API 0 CACHE STRING "Toggle unstable Zenoh-C API")
+set(Z_FEATURE_CONNECTIVITY 1 CACHE STRING "Toggle connectivity status/events API (unstable)")
```
Comment (DenisBiryukov91) [inline CMakeLists.txt:245]: We should probably also expose Z_FEATURE_CONNECTIVITY in PackageConfig.cmake.in - since it affects public API and will be handy for conditional compilation of C++ tests/examples.

File: src/api/connectivity.c [hunk 7]
```diff
@@ -0,0 +1,669 @@
...
+                                                         listener_state) == NULL) {
+        _z_session_mutex_unlock(session);
+        _z_connectivity_transport_listener_clear(listener_state);
+        z_free(listener_state);
+        ret = _Z_ERR_SYSTEM_OUT_OF_MEMORY;
+        goto exit;
+    }
+    _z_session_mutex_unlock(session);
+
+    _z_session_weak_t weak = _z_session_rc_clone_as_weak(zs);
```
Comment (DenisBiryukov91) [inline src/api/connectivity.c:553]: this call normally can not fail, since it is just an atomic variable increase

File: include/zenoh-pico/net/session.h [hunk 8]
```diff
@@ -39,6 +39,84 @@ extern "C" {
  */
 struct _z_write_filter_registration_t;
 
+#if Z_FEATURE_CONNECTIVITY == 1
+typedef struct {
+    _z_void_rc_t _callback;
```
Comment (DenisBiryukov91) [inline include/zenoh-pico/net/session.h:45]: If you store callbacks under rc and thus want to allow executing them outside of session mutex, you need to provide guarantees that callbacks will be dropped upon listener undeclaration/session close, by probably using sync_groups as introduced by https://github.com/eclipse-zenoh/zenoh-pico/pull/1158.

File: src/api/admin_space.c (hunk 5 above)
Comment (steils) [reply src/api/admin_space.c]: Fixed

File: docs/api.rst (hunk 1 above)
Comment (steils) [reply docs/api.rst]: Fixed

File: src/api/api.c [hunk 9]
```diff
@@ -896,6 +1041,206 @@ z_result_t z_info_routers_zid(const z_loaned_session_t *zs, z_moved_closure_zid_
...
+}
+
+void z_info_links_options_default(z_info_links_options_t *options) { options->transport = NULL; }
+
+z_result_t z_info_transports(const z_loaned_session_t *zs, z_moved_closure_transport_t *callback) {
+    z_result_t ret = _Z_RES_OK;
+    _z_session_t *session = _Z_RC_IN_VAL(zs);
+    _z_transport_t *transport = &session->_tp;
+    _z_transport_common_t *transport_common = _z_transport_get_common(transport);
+    if (transport_common != NULL) {
```
Comment (DenisBiryukov91) [inline src/api/api.c:1074]: This is not really thread-safe (see zp_unicast_failed)

File: src/api/connectivity.c (hunk 7 above)
Comment (steils) [reply src/api/connectivity.c:553]: Normally it's almost impossible. But it I see that `name##_rc_clone_as_weak` can return null, so why not check it?

Comment (DenisBiryukov91) [reply src/api/connectivity.c:553]: just to remove boilerplate code. Indeed it can return null, but for this (outside of bugs in rc implementation) you would require to have ~2^32 copies of rc pointer which would require roughly 32 GB of ram - you will run out of memory much earlier.

Comment (steils) [reply src/api/connectivity.c:553]: > outside of bugs in rc implementation

What if the rc implementation changes sometime? I would still keep the check here:  the cost of this check is minimal, while keeping the API usage explicit and correct.
Or if we assume that it cannot return null except for bugs, then I think it would be more correct to use an assert inside `name##_rc_clone_as_weak()`, not to remove checks at call places.

File: include/zenoh-pico/net/session.h (hunk 8 above)
Comment (steils) [reply include/zenoh-pico/net/session.h:45]: Fixed, I hope

File: src/api/api.c (hunk 9 above)
Comment (steils) [reply src/api/api.c:1074]: I hope it would it be okay to add a new _mutex_transport. If so, fixed.

File: src/api/api.c [hunk 10]
```diff
@@ -896,6 +1041,211 @@ z_result_t z_info_routers_zid(const z_loaned_session_t *zs, z_moved_closure_zid_
...
+
+    if (transport_common != NULL) {
+        _z_transport_peer_mutex_unlock(transport_common);
+    }
+    _z_session_transport_mutex_unlock(session);
+
+out:
+    z_transport_drop(opt.transport);
+
+    void *ctx = callback->_this._val.context;
```
Comment (DenisBiryukov91) [inline src/api/api.c]: why not just z_closure_link_drop() ?

File: src/api/api.c [hunk 11]
```diff
@@ -896,6 +1041,211 @@ z_result_t z_info_routers_zid(const z_loaned_session_t *zs, z_moved_closure_zid_
...
+        default:
+            break;
+    }
+
+    if (transport_common != NULL) {
+        _z_transport_peer_mutex_unlock(transport_common);
+    }
+    _z_session_transport_mutex_unlock(session);
+
+    void *ctx = callback->_this._val.context;
```
Comment (DenisBiryukov91) [inline src/api/api.c]: why not just z_closure_transport_drop() ?

File: src/api/connectivity.c [hunk 12]
```diff
@@ -0,0 +1,958 @@
...
+            (_z_connectivity_transport_cb_state_t *)callback_snapshot._val;
+        _z_connectivity_replay_transport_history(session, callback_state);
+        _z_void_rc_drop(&callback_snapshot);
+    }
+
+exit:
+    if (locked_transport_common != NULL) {
+        _z_transport_peer_mutex_unlock(locked_transport_common);
+    }
+    if (_z_sync_group_check(&callback_drop_sync_group)) {
```
Comment (DenisBiryukov91) [inline src/api/connectivity.c]: no need for this check, _z_sync_group_t is double drop safe

File: src/api/connectivity.c [hunk 13]
```diff
@@ -0,0 +1,958 @@
...
+        _z_connectivity_link_cb_state_t *callback_state = (_z_connectivity_link_cb_state_t *)callback_snapshot._val;
+        _z_connectivity_replay_link_history(session, callback_state, has_transport_filter, &transport_filter);
+        _z_void_rc_drop(&callback_snapshot);
+    }
+
+exit:
+    if (locked_transport_common != NULL) {
+        _z_transport_peer_mutex_unlock(locked_transport_common);
+    }
+    if (_z_sync_group_check(&callback_drop_sync_group)) {
```
Comment (DenisBiryukov91) [inline src/api/connectivity.c]: No need for this check, since _z_sync_group_t is double-drop safe

File: src/api/api.c (hunk 10 above)
Comment (steils) [reply src/api/api.c]: Thanks, fixed

File: src/api/api.c (hunk 11 above)
Comment (steils) [reply src/api/api.c]: Thanks, fixed

File: src/api/connectivity.c (hunk 12 above)
Comment (steils) [reply src/api/connectivity.c]: Thanks, fixed

File: src/api/connectivity.c (hunk 13 above)
Comment (steils) [reply src/api/connectivity.c]: Thanks, fixed

File: src/api/admin_space.c [hunk 14]
```diff
@@ -66,25 +74,32 @@ static z_result_t _ze_admin_space_transport_ke(z_owned_keyexpr_t *ke, const z_id
...
-
-    char buf[128];
-    int n = snprintf(buf, sizeof(buf), "%s%s%.*s%s%s%s%s%s%s%s%s%s%.*s", _Z_KEYEXPR_AT, _Z_KEYEXPR_SEPARATOR,
+    z_owned_string_t peer_zid_str;
+    _Z_CLEAN_RETURN_IF_ERR(z_id_to_string(peer_zid, &peer_zid_str), z_string_drop(z_string_move(&zid_str)));
+    z_owned_string_t link_id_str;
+    _Z_CLEAN_RETURN_IF_ERR(z_id_to_string(link_id, &link_id_str), z_string_drop(z_string_move(&zid_str));
+                           z_string_drop(z_string_move(&peer_zid_str)));
+
+    char buf[256];
```
Comment (DenisBiryukov91) [inline src/api/admin_space.c]: Shouldn't these and other fixed sizes be exposed somewhere as macros ?

Comment (steils) [reply src/api/admin_space.c]: Done

File: src/transport/multicast/lease.c [hunk 15]
```diff
@@ -148,10 +148,28 @@ void *_zp_multicast_lease_task(void *ztm_arg) {
...
+                    _z_connectivity_peer_disconnected(
+                        s, &disconnected_peer, true, mtu, is_streamed, is_reliable,
+                        _z_string_check(&disconnected_peer._link_src) ? &disconnected_peer._link_src : NULL,
+                        _z_string_check(&disconnected_peer._link_dst) ? &disconnected_peer._link_dst : NULL);
+                    _z_transport_peer_common_clear(&disconnected_peer);
+                    _z_transport_peer_mutex_lock(&ztm->_common);
+                    curr_list = ztm->_peers;
+                    prev = NULL;
+                    prev_drop = NULL;
+                    continue;
```
Comment (DenisBiryukov91) [inline src/transport/multicast/lease.c]: it does not look like this continue is needed here

File: src/transport/multicast/lease.c [hunk 16]
```diff
@@ -148,10 +148,28 @@ void *_zp_multicast_lease_task(void *ztm_arg) {
...
                     ztm->_peers = _z_transport_peer_multicast_slist_drop_element(ztm->_peers, prev_drop);
+#if Z_FEATURE_CONNECTIVITY == 1
+                    _z_transport_peer_mutex_unlock(&ztm->_common);
+                    _z_connectivity_peer_disconnected(
+                        s, &disconnected_peer, true, mtu, is_streamed, is_reliable,
+                        _z_string_check(&disconnected_peer._link_src) ? &disconnected_peer._link_src : NULL,
+                        _z_string_check(&disconnected_peer._link_dst) ? &disconnected_peer._link_dst : NULL);
+                    _z_transport_peer_common_clear(&disconnected_peer);
+                    _z_transport_peer_mutex_lock(&ztm->_common);
+                    curr_list = ztm->_peers;
```
Comment (DenisBiryukov91) [inline src/transport/multicast/lease.c]: you are resetting a list cursor here, which means that if you had previously peers that received a message, they were just flipped to not received. After cursor reset you are going to check them again and since they were just marked as not received you are going falsely drop them.

I think you would need to gather all your notifications into buffer while holding a mutex, and report them after the loop ends.

Also given that dispatching notifications might block due to the user, all the time management logic here becomes very unreliable (it sort of already was, but not to the same degree), probably it is worth starting to use z_clock_t for more reliable time measurements, not sure if this should be the subject this PR though - it is probably ok to just create an issue ...

Comment (steils) [reply src/transport/multicast/lease.c]: Thanks. Fixed

File: src/transport/multicast/lease.c [hunk 17]
```diff
@@ -148,31 +176,22 @@ void *_zp_multicast_lease_task(void *ztm_arg) {
...
                     _z_session_t *s = _z_transport_common_get_session(&ztm->_common);
 #if Z_FEATURE_CONNECTIVITY == 1
-                    _z_transport_peer_common_t disconnected_peer = {0};
-                    uint16_t mtu = 0;
-                    bool is_streamed = false;
-                    bool is_reliable = false;
-                    _z_transport_get_link_properties(&ztm->_common, &mtu, &is_streamed, &is_reliable);
-                    _z_transport_peer_common_copy(&disconnected_peer, &curr_peer->common);
+                    _z_transport_peer_multicast_slist_t *new_dropped_peers =
+                        _z_transport_peer_multicast_slist_push(dropped_peers, curr_peer);
```
Comment (DenisBiryukov91) [inline src/transport/multicast/lease.c]: copying peers might be an expensive operation (and it is also broken as of now, since it does not copy resource tree - but this is irrelevant here). Could you rather gather all the peers you are about to drop in a separate list (can probably use _z_slist_extract_filter to split  peers based on their common._received value - this way you do not even need to do any allocations), and then process drop event reporting separately.

File: src/transport/unicast/lease.c
Comment (DenisBiryukov91) [inline src/transport/unicast/lease.c]: Why use a list here, if you only have a single element ?

File: src/transport/unicast/lease.c [hunk 18]
```diff
@@ -171,31 +199,22 @@ void *_zp_unicast_lease_task(void *ztu_arg) {
                     if (drop_peer) {
                         _z_session_t *zs = _z_transport_common_get_session(&ztu->_common);
 #if Z_FEATURE_CONNECTIVITY == 1
-                        _z_transport_peer_common_t disconnected_peer = {0};
-                        uint16_t mtu = 0;
-                        bool is_streamed = false;
-                        bool is_reliable = false;
-                        _z_transport_get_link_properties(&ztu->_common, &mtu, &is_streamed, &is_reliable);
-                        _z_transport_peer_common_copy(&disconnected_peer, &curr_peer->common);
+                        _z_transport_peer_unicast_slist_t *new_dropped_peers =
```
Comment (DenisBiryukov91) [inline src/transport/unicast/lease.c]: same comment as for multicast peers

File: src/transport/unicast/lease.c
Comment (steils) [reply src/transport/unicast/lease.c]: Because `ztu->_peers` is always a slist in both client and peer modes.
In client mode it is effectively a single active peer. But in peer mode it can have multiple accepted connections.

File: src/transport/unicast/lease.c (hunk 18 above)
Comment (steils) [reply src/transport/unicast/lease.c]: Done.

File: src/transport/multicast/rx.c [hunk 19]
```diff
@@ -352,8 +421,25 @@ static z_result_t _z_multicast_handle_join_inner(_z_transport_multicast_t *ztm,
         // Check representing capabilities
         if ((msg->_seq_num_res != Z_SN_RESOLUTION) || (msg->_req_id_res != Z_REQ_RESOLUTION) ||
             (msg->_batch_size != Z_BATCH_MULTICAST_SIZE)) {
+#if Z_FEATURE_CONNECTIVITY == 1
+            _z_transport_peer_common_t disconnected_peer = {0};
+            uint16_t mtu = 0;
+            bool is_streamed = false;
+            bool is_reliable = false;
+            _z_transport_get_link_properties(&ztm->_common, &mtu, &is_streamed, &is_reliable);
+            _z_transport_peer_common_copy(&disconnected_peer, &entry->common);
```
Comment (DenisBiryukov91) [inline src/transport/multicast/rx.c]: do we really need to make this copy (which might also include copy of the wbuf) to report an event ?

Comment (steils) [reply src/transport/multicast/rx.c]: I think you are right. We need only certain fields, not the entire structure. Fixed.

File: src/api/admin_space.c [hunk 20]
```diff
@@ -681,21 +833,82 @@ static void _ze_admin_space_query_handle_pico(const z_loaned_query_t *query, _z_
...
-        _Z_ERROR("Dropped admin space query - session closed");
+#if Z_FEATURE_CONNECTIVITY == 1
+static void _ze_admin_space_query_handle_connectivity_session(const z_loaned_query_t *query, _z_session_t *session,
+                                                              _ze_admin_space_reply_list_t **replies) {
+    _z_session_transport_mutex_lock(session);
+    if (session->_tp._type != _Z_TRANSPORT_UNICAST_TYPE) {
+        _z_session_transport_mutex_unlock(session);
         return;
     }
+#if Z_FEATURE_MULTI_THREAD == 1
```
Comment (DenisBiryukov91) [inline src/api/admin_space.c]: Z_FEATURE_MULTI_THREAD guard is not needed here and in other places, since we have a stub version of _z_transport_peer_mutex_lock/unlock for single threaded case

File: src/api/admin_space.c [hunk 21]
```diff
@@ -720,67 +933,531 @@ static void _ze_admin_space_query_handler(z_loaned_query_t *query, void *ctx) {
...
+        session->_admin_space_queryable_id = 0;
+    } else if (ret == _Z_RES_OK) {
+        ret = undeclare_queryable_ret;
+    }
+}
+    return ret;
+#endif
 }
 
 z_result_t zp_stop_admin_space(z_loaned_session_t *zs) {
```
Comment (DenisBiryukov91) [inline src/api/admin_space.c:1356]: This function and zp_start_admin_space is not thread safe (not that they were before, though), because unprotected reading a modifying non-atomic _admin_space_queryable_id, _admin_space_session_queryable_id, etc at the same time is undefined behavior.

File: src/api/connectivity.c [hunk 22]
```diff
@@ -0,0 +1,874 @@
...
+    }
+
+    _z_connectivity_transport_listener_intmap_clear(&snapshot);
+}
+
+static void _z_connectivity_dispatch_link_event(_z_session_t *session, _z_info_link_event_t *event, bool is_multicast) {
+    _z_connectivity_link_listener_intmap_t snapshot = _z_connectivity_link_listener_intmap_make();
+
+    _z_session_mutex_lock(session);
+    snapshot = _z_connectivity_link_listener_intmap_clone(&session->_connectivity_link_event_listeners);
```
Comment (DenisBiryukov91) [inline src/api/connectivity.c:283]: I'm wondering, maybe it is better to just put this map under a separate mutex and hold it while dispatching events, given that link events are pretty rare, and thus should not introduce a lot of latency? This would also simplify synchronization of callbacks drop on undeclare.

File: src/session/utils.c [hunk 23]
```diff
@@ -174,17 +215,35 @@ void _z_session_clear(_z_session_t *zn) {
 #if Z_FEATURE_AUTO_RECONNECT == 1
     _z_config_clear(&zn->_config);
 #endif
-    _z_transport_clear(&zn->_tp);
+    _z_session_transport_mutex_lock(zn);
```
Comment (DenisBiryukov91) [inline src/session/utils.c:234]: why not just call _z_transport_clear while holding the mutex ?

File: src/api/connectivity.c (hunk 22 above)
Comment (steils) [reply src/api/connectivity.c:283]: As discussed, we'll make it a separate task.

File: src/session/utils.c (hunk 23 above)
Comment (steils) [reply src/session/utils.c:234]: Just some optimization. It would unnecessarily extend the critical section over task joins and transport cleanup. Also after the detach, generic `_z_transport_clear()` can no longer be used anyway, since it dispatches by zt->_type.

File: src/api/admin_space.c (hunk 20 above)
Comment (steils) [reply src/api/admin_space.c]: Fixed

File: src/api/admin_space.c (hunk 21 above)
Comment (steils) [reply src/api/admin_space.c:1356]: Fixed

File: src/session/utils.c (hunk 23 above)
Comment (DenisBiryukov91) [reply src/session/utils.c:234]: I do not think we care much about extending critical section here, because we are closing session, so very likely terminating the application - so it is not in the hot path. I believe having a copy-pasted piece of code is way worse here.  Detach issue will need to be addressed separately though.

Comment (steils) [reply src/session/utils.c:234]: Ok, let's address the detach issue separately.
Fixed, cleaned up the code.

File: .github/workflows/integration.yaml [hunk 24]
```diff
@@ -76,7 +76,7 @@ jobs:
 
       - name: Test debug
         run: make test
-        timeout-minutes: 30
+        timeout-minutes: 40
```
Comment (gmartin82) [inline .github/workflows/integration.yaml:79]: While not related to this particular ticket, I think we really need to stop increasing this and fix the underlying issue (at some point, not on this ticket) by building zenohd separately and potentially splitting up the tests.

File: src/api/admin_space.c [hunk 25]
```diff
@@ -152,6 +155,76 @@ static z_result_t _ze_admin_space_pico_transports_0_peers_peer_ke(z_owned_keyexp
...
+                                                       const z_id_t *peer_zid) {
+    z_internal_keyexpr_null(ke);
+
+    z_owned_string_t zid_str;
+    _Z_RETURN_IF_ERR(z_id_to_string(zid, &zid_str));
+    z_owned_string_t peer_zid_str;
+    _Z_CLEAN_RETURN_IF_ERR(z_id_to_string(peer_zid, &peer_zid_str), z_string_drop(z_string_move(&zid_str)));
+
+    char buf[_ZE_ADMIN_SPACE_SESSION_TRANSPORT_KE_BUF_LEN];
+    int n = snprintf(buf, sizeof(buf), "%s%s%.*s%s%s%s%s%s%.*s", _Z_KEYEXPR_AT, _Z_KEYEXPR_SEPARATOR,
```
Comment (gmartin82) [inline src/api/admin_space.c]: Why not use `_ze_admin_space_build_ke()`?

File: src/api/admin_space.c [hunk 26]
```diff
@@ -152,6 +155,76 @@ static z_result_t _ze_admin_space_pico_transports_0_peers_peer_ke(z_owned_keyexp
...
+    z_owned_string_t zid_str;
+    _Z_RETURN_IF_ERR(z_id_to_string(zid, &zid_str));
+    z_owned_string_t peer_zid_str;
+    _Z_CLEAN_RETURN_IF_ERR(z_id_to_string(peer_zid, &peer_zid_str), z_string_drop(z_string_move(&zid_str)));
+    z_owned_string_t link_id_str;
+    _Z_CLEAN_RETURN_IF_ERR(z_id_to_string(link_id, &link_id_str), z_string_drop(z_string_move(&zid_str));
+                           z_string_drop(z_string_move(&peer_zid_str)));
+
+    char buf[_ZE_ADMIN_SPACE_SESSION_LINK_KE_BUF_LEN];
+    int n = snprintf(buf, sizeof(buf), "%s%s%.*s%s%s%s%s%s%.*s%s%s%s%.*s", _Z_KEYEXPR_AT, _Z_KEYEXPR_SEPARATOR,
```
Comment (gmartin82) [inline src/api/admin_space.c]: Why not use `_ze_admin_space_build_ke()`?

File: src/api/admin_space.c [hunk 27]
```diff
@@ -720,67 +929,534 @@ static void _ze_admin_space_query_handler(z_loaned_query_t *query, void *ctx) {
...
+    if (admin_space_queryable_id != 0) {
+        z_result_t queryable_ret = _ze_admin_space_undeclare_queryable(zs, admin_space_queryable_id);
+        if (queryable_ret == _Z_RES_OK) {
+            session->_admin_space_queryable_id = 0;
+        } else if (ret == _Z_RES_OK) {
+            ret = queryable_ret;
+        }
     }
+
+out:
```
Comment (gmartin82) [inline src/api/admin_space.c:1423]: I have some concerns over the introduction of using goto's in the code base. This isn't something we normally use and is discouraged by MISRA C.

File: src/api/admin_space.c [hunk 28]
```diff
@@ -720,67 +929,534 @@ static void _ze_admin_space_query_handler(z_loaned_query_t *query, void *ctx) {
...
+err_queryable: {
+    z_result_t undeclare_queryable_ret = _ze_admin_space_undeclare_queryable(zs, session->_admin_space_queryable_id);
+    if (undeclare_queryable_ret == _Z_RES_OK) {
+        session->_admin_space_queryable_id = 0;
+    } else if (ret == _Z_RES_OK) {
+        ret = undeclare_queryable_ret;
+    }
+}
+#endif
+out:
```
Comment (gmartin82) [inline src/api/admin_space.c:1351]: Same comment re: gotos in code base.

File: src/api/admin_space.c [hunk 29]
```diff
@@ -720,67 +929,534 @@ static void _ze_admin_space_query_handler(z_loaned_query_t *query, void *ctx) {
...
+    }
+
+    ret = _ze_admin_space_session_queryable_ke(&ke, &zid);
+    if (ret != _Z_RES_OK) {
+        _z_session_weak_drop(session_session_weak);
+        z_free(session_session_weak);
+        goto err_queryable;
+    }
+
+    ret = _ze_admin_space_declare_session_queryable(zs, &session->_admin_space_session_queryable_id,
```
Comment (gmartin82) [inline src/api/admin_space.c:1252]: Why are we declaring two queryables in different ways here?

File: src/api/admin_space.c [hunk 30]
```diff
@@ -310,6 +437,31 @@ static z_result_t _ze_admin_space_add_reply(_z_json_encoder_t *je, const z_loane
     return _Z_RES_OK;
 }
 
+#if Z_FEATURE_CONNECTIVITY == 1
+static z_result_t _ze_admin_space_add_reply_payload(const z_loaned_keyexpr_t *ke, z_moved_bytes_t *payload,
```
Comment (gmartin82) [inline src/api/admin_space.c]: Why are we introducing a second mechanism to add a reply here instead of keeping things consistent?

**Review by gmartin82 — CHANGES_REQUESTED** [review]
I have some concerns about style inconsistencies here that should be cleaned up one way or another.

I’m also concerned about the introduction of `goto`, which the existing codebase has generally avoided to date. While `goto` can sometimes make control flow cleaner in C (and I think it does in this case), MISRA C:2023 (e.g. rule 15.1) discourages its use, and I do not think we should start introducing it without an explicit decision on whether we want to allow that pattern going forward. At the moment this feels inconsistent with the current style of the codebase.

File: .github/workflows/integration.yaml (hunk 24 above)
Comment (steils) [reply .github/workflows/integration.yaml:79]: I agree, but let's work on it in a separate issue.

File: src/api/admin_space.c (hunk 25 above)
Comment (steils) [reply src/api/admin_space.c]: I can use it, but we'll have to increase _ZE_ADMIN_SPACE_KE_BUF_LEN from 128 to 256, so the connectivity keyexprs would fit.

File: src/api/admin_space.c (hunk 27 above)
Comment (steils) [reply src/api/admin_space.c:1423]: It's a usual C-style rollback path, to avoid duplicating lots of code on error paths, and improves readability in general.
MISRA C generally discourages goto, but doesn't ban it. Structured, forward-only jumps to a common cleanup/rollback label in the same function is a usual pattern in C. See MISRA C:2023 15.2 and later, in particular 15.5 (having the single point of exit is why we use goto here).

File: src/api/admin_space.c (hunk 28 above)
Comment (steils) [reply src/api/admin_space.c:1351]: Replied there

File: src/api/admin_space.c (hunk 29 above)
Comment (steils) [reply src/api/admin_space.c:1252]: Because the two queryables have different locality requirements.
The `@/<zid>/pico/**` queryable is a normal admin-space queryable, so z_declare_queryable() is sufficient.
The `@/<zid>/session/**` queryable must be **SessionLocal** according to the RFC [https://github.com/eclipse-zenoh/roadmap/blob/main/rfcs/ALL/Connectivity%20Status%20and%20Events.md]. For that I call `_z_register_queryable(..., Z_LOCALITY_SESSION_LOCAL, ...)` in `_ze_admin_space_declare_session_queryable()`.

File: src/api/admin_space.c (hunk 25 above)
Comment (steils) [reply src/api/admin_space.c]: Done.

File: src/api/admin_space.c (hunk 26 above)
Comment (steils) [reply src/api/admin_space.c]: Done.

File: src/api/admin_space.c (hunk 30 above)
Comment (steils) [reply src/api/admin_space.c]: Thanks. Fixed.

File: src/api/admin_space.c (hunk 27 above)
Comment (gmartin82) [reply src/api/admin_space.c:1423]: My point is you’re the only person using gotos in this code base. It needs wider discussion.
